tools/gfx-unit-test/root-shader-parameter.cpp

```diff
@@ -0,0 +1,135 @@
+        Slang::List<ComPtr<IResourceView>> srvs, uavs;
+
+        for (uint32_t i = 0; i < 9; i++)
+        {
+            buffers.add(createBuffer(device, i == 0 ? 10 : i));
+
+            ComPtr<IResourceView> bufferView;
+            IResourceView::Desc viewDesc = {};
+            viewDesc.type = IResourceView::Type::UnorderedAccess;
+            viewDesc.format = Format::Unknown;
+            GFX_CHECK_CALL_ABORT(
+                device->createBufferView(buffers[i], nullptr, viewDesc, bufferView.writeRef()));
+            uavs.add(bufferView);
+
+            viewDesc.type = IResourceView::Type::ShaderResource;
+            viewDesc.format = Format::Unknown;
+            GFX_CHECK_CALL_ABORT(
+                device->createBufferView(buffers[i], nullptr, viewDesc, bufferView.writeRef()));
+            srvs.add(bufferView);
+        }
+
+        ComPtr<IShaderObject> rootObject;
+        device->createMutableRootShaderObject(shaderProgram, rootObject.writeRef());
+
+        ComPtr<IShaderObject> g, s1, s2;
+        device->createMutableShaderObject(
+            slangReflection->findTypeByName("S0"), ShaderObjectContainerType::None, g.writeRef());
+        device->createMutableShaderObject(
+            slangReflection->findTypeByName("S1"), ShaderObjectContainerType::None, s1.writeRef());
+        device->createMutableShaderObject(
+            slangReflection->findTypeByName("S1"), ShaderObjectContainerType::None, s2.writeRef());
+
+        {
+            auto cursor = ShaderCursor(s1);
+            cursor["c0"].setResource(srvs[2]);
+            cursor["c1"].setResource(uavs[3]);
+            cursor["c2"].setResource(srvs[4]);
+        }
+        {
+            auto cursor = ShaderCursor(s2);
+            cursor["c0"].setResource(srvs[5]);
+            cursor["c1"].setResource(uavs[6]);
+            cursor["c2"].setResource(srvs[7]);
+        }
+        {
+            auto cursor = ShaderCursor(g);
+            cursor["b0"].setResource(srvs[0]);
+            cursor["b1"].setResource(srvs[1]);
+            cursor["s1"].setObject(s1);
+            cursor["s2"].setObject(s2);
+        }
+        {
+            auto cursor = ShaderCursor(rootObject);
+            cursor["g"].setObject(g);
+            cursor["buffer"].setResource(uavs[8]);
+        }
+
+        {
+            ICommandQueue::Desc queueDesc = { ICommandQueue::QueueType::Graphics };
+            auto queue = device->createCommandQueue(queueDesc);
+
+            auto commandBuffer = transientHeap->createCommandBuffer();
+            {
+                auto encoder = commandBuffer->encodeComputeCommands();
+                auto root = encoder->bindPipeline(pipelineState);
+                root->copyFrom(rootObject, transientHeap);
+                encoder->dispatchCompute(1, 1, 1);
+                encoder->endEncoding();
+            }
+
+            commandBuffer->close();
+            queue->executeCommandBuffer(commandBuffer);
```

Comment:
In terms of further convenience stuff, it seems like we will want some RAII-style types to deal with command buffers and encoders.

At the very least, it would be good for a user to be able to do something like (ignoring naming):
```
{
    CommandBufferRAII commandBuffer(queue, transientHeap);
    ComputeEncoderRAII encoder(commandBuffer);
    ...
    encoder.dispatchCompute(1, 1, 1);
    // end
}
```
The destructor for the RAII encoder would then automatically close it, and the destructor for the RAII command buffer would automatically close *and* execute it on the provided queue.

For common cases used by simple apps, the RAII encoder types should then also support a convenience constructor that works directly with a queue:

```
{
    ComputeEncoderRAII encoder(queue, transientHeap);
    ...
    encoder.dispatchCompute(1, 1, 1);
}
```

That variant should have the encoder automatically manage its own constant buffer and close/execute the buffer when the RAII encoder is destructed.
// end